discord/rate_limiting.py

```diff
@@ -0,0 +1,558 @@
+        self._buckets: dict[str, Bucket] = {}
+        self.ready = True
+        self.global_concurrency = GlobalRateLimit(concurrency, per)
+
+        gc.callbacks.append(self._collect_buckets)
+
+        # temporary buckets, different from normal "permanent" buckets
+        # which cannot be tracked via remaining or reset_after in headers
+        self._temp_buckets: dict[str, DynamicBucket] = {}
+
+    def _collect_buckets(
+        self, phase: Literal["start", "stop"], info: dict[str, int]
+    ) -> None:
+        del info
+
+        if phase == "stop":
+            return
+
+        for id, bucket in self._buckets.copy().items():
+            if bucket.garbage:
+                del self._buckets[id]
+
+    async def close(self) -> None:
+        gc.callbacks.remove(self._collect_buckets)
+
+    async def append(self, id: str, bucket: Bucket) -> None:
+        """Append a permanent bucket.
+
+        Parameters
+        ----------
+        id: :class:`str`
+            This bucket's identifier.
+        bucket: :class:`.Bucket`
+            The bucket to append.
+        """
+
+        self._buckets[id] = bucket
+
+    async def get(self, id: str) -> Bucket | None:
+        """Get a permanent bucket.
+
+        Parameters
+        ----------
+        id: :class:`str`
+            This bucket's identifier.
+
+        Returns
+        -------
+        Optional[:class:`.Bucket`]
+        """
+
+        return self._buckets.get(id)
+
+    async def get_or_create(self, id: str) -> Bucket:
+        """Get or create a permanent bucket.
+
+        Parameters
+        ----------
+        id: :class:`str`
+            This bucket's identifier.
+
+        Returns
+        -------
+        :class:`.Bucket`
+        """
+
+        buc = await self.get(id)
+
+        if not buc:
+            buc = Bucket()
+            await self.append(id, buc)
+
+        return buc
+
+    async def temp_bucket(self, id: str) -> DynamicBucket | None:
+        """Fetch a temporary bucket.
+
+        Parameters
+        ----------
+        id: :class:`str`
+            This bucket's identifier.
+
+        Returns
+        -------
+        Optional[:class:`.DynamicBucket`]
+        """
+
+        return self._temp_buckets.get(id)
+
+    async def push_temp_bucket(self, id: str, bucket: DynamicBucket) -> None:
+        """Push a temporary bucket to storage.
+
+        Parameters
+        ----------
+        id: :class:`str`
+            This bucket's identifier.
+        """
+
+        self._temp_buckets[id] = bucket
+
+    async def delete_temp_bucket(self, id: str) -> None:
+        """Delete a temporary bucket which may be in storage.
+
+        Parameters
+        ----------
+        id: :class:`str`
+            This bucket's identifier.
+        """
+
+        if id in self._temp_buckets:
+            del self._temp_buckets[id]
+
+
+class DynamicBucket:
+    """A dynamic bucket for on-the-fly rate limits. Should not be used inside a bot directly!"""
```

Comment:
Could you make the warning more explanatory and maybe put it in a `.. warning::` block?